relay-metrics/src/protocol.rs

```diff
@@ -0,0 +1,796 @@
+    } else if let Ok(float) = string.parse::<f64>() {
+        if float < 0f64 {
+            None
+        } else {
+            Some(UnixTimestamp::from_secs(float.trunc() as u64))
+        }
+    } else {
+        None
+    }
+}
+
+/// A single metric value representing the payload sent from clients.
+///
+/// As opposed to bucketed metric aggregations, this single metrics always represent a single
+/// submission and cannot store multiple values.
+///
+/// See the [crate documentation](crate) for general information on Metrics.
+///
+/// # Submission Protocol
+///
+/// ```text
+/// <name>[@unit]:<value>|<type>|'<timestamp>|#<tag_key>:<tag_value>,<tag>
+/// ```
+///
+/// See the field documentation on this struct for more information on the components. An example
+/// submission looks like this:
+///
+/// ```text
+/// endpoint.response_time@ms:57|h|'1615889449|#route:user_index
+/// endpoint.hits:1|c|'1615889449|#route:user_index
+/// ```
+///
+/// To parse a submission payload, use [`Metric::parse_all`].
+///
+/// # JSON representation
+///
+/// In addition to the submission protocol, metrics can be represented as structured data in JSON.
+/// Field values are the same with a single exception: The timestamp is required in JSON notation.
+///
+/// ```json
+/// {
+///   "name": "endpoint.response_time",
+///   "unit": "ms",
+///   "value": 57,
+///   "type": "h",
+///   "timestamp": 1615889449,
+///   "tags": {
+///     "route": "user_index"
+///   }
+/// }
+/// ```
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct Metric {
+    /// The name of the metric without its unit.
+    ///
+    /// Metric names can consist of ASCII alphanumerics, underscores and periods.
+    pub name: String,
+    /// The unit of the metric value.
+    ///
+    /// Units augment metric values by giving them a magnitude and semantics. There are certain
+    /// types of units that are subdivided in their precision, such as the [`DurationPrecision`] for
+    /// time measurements.
+    ///
+    /// The unit can be omitted and defaults to [`MetricUnit::None`].
+    #[serde(default, skip_serializing_if = "MetricUnit::is_none")]
+    pub unit: MetricUnit,
+    /// The value of the metric.
+    ///
+    /// [Histograms](MetricType::Histogram) and [counters](MetricType::Counter) require numeric
+    /// values which can either be integral or floating point. In contrast, [sets](MetricType::Set)
+    /// and [gauges](MetricType::Gauge) can store any unique value including custom strings.
+    pub value: MetricValue,
+    /// The type of the metric, determining its aggregation and evaluation.
+    #[serde(rename = "type")]
+    pub ty: MetricType,
+    /// The timestamp for this metric value.
+    ///
+    /// In the SDK protocol, timestamps are optional and preceded with a single quote `'`. Supply a
+    /// default timestamp to [`Metric::parse`] or [`Metric::parse_all`] to insert a default
+    /// timestamp.
+    pub timestamp: UnixTimestamp,
```

Comment:
would it not make more sense to not associate timestamps with metrics at all and put them into a countainer holding the timestamp? This would simplify a lot here (particularly also the currently quite weird parsing) and would allow the metric type to be used a bit more flexibly.

Reply:
Yes, that might actually make sense. There are really two cases we can consider:

1. A series of values reported for the *same* metric with different timestamps
2. Multiple metrics for *different* metrics at one timestamp

Going back, case 2 is probably more likely to occur for client submission, so we should optimize for that. I'm wondering where we'd put it in the protocol, then, since it doesn't make sense to keep it in this spot. Any suggestions?

Reply:
`{"timestamp": 1615889449, "metrics": ["endpoint.hits:1|c|#route:user_index", ...]}\n`

Reply:
Will revisit this with bucketing in the follow-up.